File: Algorithm/Python/201. Bitwise AND of Numbers Range.py
# -*- coding: utf-8 -*-
# @Author: LC
# @Date:   2016-07-20 15:41:02
# @Last modified by:   WuLC
# @Last Modified time: 2016-07-20 15:41:42
# @Email: liangchaowu5@gmail.com

# find the common prefix of the two numbers
class Solution(object):
    def rangeBitwiseAnd(self, m, n):
        """
        :type m: int
        :type n: int
        :rtype: int
        """
        count = 0
        while m != n:
            m >>= 1
            n >>= 1
            count += 1
        for i in xrange(count):
            m <<= 1
        return m

File: Algorithm/Java/201. Bitwise AND of Numbers Range.java
/*
* @Author: WuLC
* @Date:   2017-07-14 08:54:58
* @Last Modified by:   WuLC
* @Last Modified time: 2017-07-14 21:42:03
* @Email: liangchaowu5@gmail.com
*/


// referer: https://discuss.leetcode.com/topic/12133/bit-operation-solution-java
// if m != n, the rightmost bit must be 0
public class Solution 
{
    public int rangeBitwiseAnd(int m, int n) 
    {
        int  count = 1;
        while(m != n)
        {
            m >>= 1;
            n >>= 1;
            count <<= 1;
        }
        return m * count;
    }
}